lib/web/desktop.go

```diff
@@ -117,30 +113,110 @@ func createDesktopConnection(
 		return trace.WrapWithMessage(err, "failed to connect to windows_desktop_service at %q: %v", service.GetAddr(), err)
 	}
 	defer serviceCon.Close()
-	tlsConfig := ctx.clt.Config()
-	// Pass target desktop UUID via SNI.
-	tlsConfig.ServerName = desktopName + desktop.SNISuffix
-	serviceConTLS := tls.Client(serviceCon, ctx.clt.Config())
-	log.Debug("Connected to windows_desktop_service")
 
-	tdpConn := tdp.NewConn(serviceConTLS)
-	err = tdpConn.OutputMessage(tdp.ClientUsername{Username: username})
-	if err != nil {
-		return trace.Wrap(err)
-	}
-	err = tdpConn.OutputMessage(tdp.ClientScreenSpec{Width: uint32(width), Height: uint32(height)})
+	pc, err := proxyClient(r.Context(), ctx, h.ProxyHostPort())
 	if err != nil {
 		return trace.Wrap(err)
 	}
+	defer pc.Close()
 
 	websocket.Handler(func(conn *websocket.Conn) {
+		tlsConfig, err := desktopTLSConfig(r.Context(), conn, pc, ctx, desktopName, username, site.GetName())
+		if err != nil {
+			writeError(err, conn, log)
+			return
+		}
+		serviceConTLS := tls.Client(serviceCon, tlsConfig)
+
+		if err := serviceConTLS.Handshake(); err != nil {
+			writeError(err, conn, log)
+			return
+		}
+		log.Debug("Connected to windows_desktop_service")
+		tdpConn := tdp.NewConn(serviceConTLS)
+		err = tdpConn.OutputMessage(tdp.ClientUsername{Username: username})
+		if err != nil {
+			writeError(err, conn, log)
+			return
+		}
+		err = tdpConn.OutputMessage(tdp.ClientScreenSpec{Width: uint32(width), Height: uint32(height)})
+		if err != nil {
+			writeError(err, conn, log)
+			return
+		}
 		if err := proxyWebsocketConn(conn, serviceConTLS); err != nil {
 			log.WithError(err).Warningf("Error proxying a desktop protocol websocket to windows_desktop_service")
 		}
 	}).ServeHTTP(w, r)
 	return nil
 }
 
+func writeError(err error, ws *websocket.Conn, log *logrus.Entry) {
+	rdpConn := tdp.NewConn(ws)
+	oerr := rdpConn.OutputMessage(tdp.Error{Message: err.Error()})
```

Comment:
```suggestion
	tdpConn := tdp.NewConn(ws)
	oerr := tdpConn.OutputMessage(tdp.Error{Message: err.Error()})
```

Reply:
`*tdp.Conn` now has a `SendError(string)` method, so you can just do:

```
tdp.NewConn(ws).SendError(err.Error())
```